File: src/LTAdmin/Repositories/EvaluationRepository.cs
using System.Data;
using LTAdmin.Data;
using LTAdmin.Models.Dto;
using LTAdmin.Models.Entities;

namespace LTAdmin.Repositories;

/// <summary>Dépôt des évaluations et des notes de contrôle continu.</summary>
public sealed class EvaluationRepository : RepositoryBase
{
    public EvaluationRepository(AccessDatabase database) : base(database)
    {
    }

    // ----- PERIODE_EVAL -----

    public PeriodeEval? GetPeriode(int idPeriode)
        => QuerySingle(
            $"SELECT * FROM {Q(Tables.PeriodeEval)} WHERE {Q("ID_PERIODE")} = ?",
            MapPeriode, P(idPeriode));

    public List<PeriodeEval> ListPeriodes(int? idAnnee = null)
        => idAnnee.HasValue
            ? QueryList($"SELECT * FROM {Q(Tables.PeriodeEval)} WHERE {Q("ID_ANNEE")} = ? ORDER BY {Q("ORDRE_PER")}", MapPeriode, P(idAnnee.Value))
            : QueryList($"SELECT * FROM {Q(Tables.PeriodeEval)} ORDER BY {Q("ID_ANNEE")}, {Q("ORDRE_PER")}", MapPeriode);

    public int InsertPeriode(PeriodeEval periode)
        => InsertAndGetId(
            $"INSERT INTO {Q(Tables.PeriodeEval)} ({Q("ID_ANNEE")}, {Q("CODE_PERIODE")}, {Q("LIBELLE")}, {Q("ORDRE_PER")}, " +
            $"{Q("PONDERATION")}, {Q("DATE_DEBUT")}, {Q("DATE_FIN")}, {Q("CLOTUREE")}) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
            P(periode.IdAnnee), P(periode.CodePeriode), P(periode.Libelle), P(periode.OrdrePer),
            P(periode.Ponderation), P(periode.DateDebut), P(periode.DateFin), P(periode.Cloturee ?? false));

    public int UpdatePeriode(PeriodeEval periode)
        => Execute(
            $"UPDATE {Q(Tables.PeriodeEval)} SET {Q("ID_ANNEE")} = ?, {Q("CODE_PERIODE")} = ?, {Q("LIBELLE")} = ?, {Q("ORDRE_PER")} = ?, " +
            $"{Q("PONDERATION")} = ?, {Q("DATE_DEBUT")} = ?, {Q("DATE_FIN")} = ?, {Q("CLOTUREE")} = ? WHERE {Q("ID_PERIODE")} = ?",
            P(periode.IdAnnee), P(periode.CodePeriode), P(periode.Libelle), P(periode.OrdrePer),
            P(periode.Ponderation), P(periode.DateDebut), P(periode.DateFin), P(periode.Cloturee ?? false),
            P(periode.IdPeriode));

    public int SetPeriodeCloturee(int idPeriode, bool cloturee)
        => Execute(
            $"UPDATE {Q(Tables.PeriodeEval)} SET {Q("CLOTUREE")} = ? WHERE {Q("ID_PERIODE")} = ?",
            P(cloturee), P(idPeriode));

    public int DeletePeriode(int idPeriode)
        => Execute($"DELETE FROM {Q(Tables.PeriodeEval)} WHERE {Q("ID_PERIODE")} = ?", P(idPeriode));

    // ----- EVALUATION -----

    public Evaluation? GetEvaluation(int idEvaluation)
        => QuerySingle(
            $"SELECT * FROM {Q(Tables.Evaluation)} WHERE {Q("ID_EVALUATION")} = ?",
            MapEvaluation, P(idEvaluation));

    public EvaluationDetail? GetEvaluationDetail(int idEvaluation)
        => QuerySingle(EvaluationDetailSql() + $" WHERE EV.{Q("ID_EVALUATION")} = ?", MapEvaluationDetail, P(idEvaluation));

    public List<EvaluationDetail> ListEvaluations(int? idPeriode = null, int? idClasse = null, int? idProg = null)
    {
        var conditions = new List<string>();
        var parameters = new List<System.Data.OleDb.OleDbParameter>();
        if (idPeriode.HasValue) { conditions.Add($"EV.{Q("ID_PERIODE")} = ?"); parameters.Add(P(idPeriode.Value)); }
        if (idClasse.HasValue) { conditions.Add($"P.{Q("ID_CLASSE")} = ?"); parameters.Add(P(idClasse.Value)); }
        if (idProg.HasValue) { conditions.Add($"EV.{Q("ID_PROG")} = ?"); parameters.Add(P(idProg.Value)); }
        var sql = EvaluationDetailSql();
        if (conditions.Count > 0) sql += " WHERE " + string.Join(" AND ", conditions);
        sql += $" ORDER BY EV.{Q("DATE_EVAL")}, EV.{Q("ID_EVALUATION")}";
        return QueryList(sql, MapEvaluationDetail, parameters.ToArray());
    }

    private static string EvaluationDetailSql()
        => $"SELECT EV.*, PE.{Q("LIBELLE")} AS PERIODE_LIB, PE.{Q("CLOTUREE")} AS PERIODE_CLOTUREE, " +
           $"P.{Q("ID_CLASSE")} AS ID_CLASSE, C.{Q("LIBELLE")} AS CLASSE_LIB, " +
           $"P.{Q("CODE_MATIERE")} AS CODE_MATIERE, M.{Q("MATIERE")} AS MATIERE_LIB " +
           $"FROM (((({Q(Tables.Evaluation)} AS EV " +
           $"INNER JOIN {Q(Tables.PeriodeEval)} AS PE ON EV.{Q("ID_PERIODE")} = PE.{Q("ID_PERIODE")}) " +
           $"INNER JOIN {Q(Tables.Programme)} AS P ON EV.{Q("ID_PROG")} = P.{Q("ID_PROG")}) " +
           $"INNER JOIN {Q(Tables.Classe)} AS C ON P.{Q("ID_CLASSE")} = C.{Q("ID_CLASSE")}) " +
           $"INNER JOIN {Q(Tables.Matiere)} AS M ON P.{Q("CODE_MATIERE")} = M.{Q("CODE_MATIERE")})";

    public int InsertEvaluation(Evaluation evaluation)
        => InsertAndGetId(
            $"INSERT INTO {Q(Tables.Evaluation)} ({Q("ID_PERIODE")}, {Q("ID_PROG")}, {Q("INTITULE")}, {Q("NATURE")}, " +
            $"{Q("DATE_EVAL")}, {Q("BAREME")}, {Q("POIDS")}, {Q("PUBLIEE")}) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
            P(evaluation.IdPeriode), P(evaluation.IdProg), P(evaluation.Intitule), P(evaluation.Nature),
            P(evaluation.DateEval), P(evaluation.Bareme), P(evaluation.Poids), P(evaluation.Publiee ?? false));

    public int UpdateEvaluation(Evaluation evaluation)
        => Execute(
            $"UPDATE {Q(Tables.Evaluation)} SET {Q("ID_PERIODE")} = ?, {Q("ID_PROG")} = ?, {Q("INTITULE")} = ?, {Q("NATURE")} = ?, " +
            $"{Q("DATE_EVAL")} = ?, {Q("BAREME")} = ?, {Q("POIDS")} = ?, {Q("PUBLIEE")} = ? WHERE {Q("ID_EVALUATION")} = ?",
            P(evaluation.IdPeriode), P(evaluation.IdProg), P(evaluation.Intitule), P(evaluation.Nature),
            P(evaluation.DateEval), P(evaluation.Bareme), P(evaluation.Poids), P(evaluation.Publiee ?? false),
            P(evaluation.IdEvaluation));

    public int SetEvaluationPubliee(int idEvaluation, bool publiee)
        => Execute(
            $"UPDATE {Q(Tables.Evaluation)} SET {Q("PUBLIEE")} = ? WHERE {Q("ID_EVALUATION")} = ?",
            P(publiee), P(idEvaluation));

    public int DeleteEvaluation(int idEvaluation)
        => Execute($"DELETE FROM {Q(Tables.Evaluation)} WHERE {Q("ID_EVALUATION")} = ?", P(idEvaluation));

    // ----- NOTE -----

    public Note? GetNote(int idEvaluation, int idInscription)
        => QuerySingle(
            $"SELECT * FROM {Q(Tables.Note)} WHERE {Q("ID_EVALUATION")} = ? AND {Q("ID_INSCRIPTION")} = ?",
            MapNote, P(idEvaluation), P(idInscription));

    /// <summary>Grille de saisie : tous les inscrits de la classe + leur note éventuelle.</summary>
    public List<NoteSaisieRow> ListGrilleSaisie(int idEvaluation, int idClasse)
        => QueryList(
            $"SELECT N.{Q("ID_NOTE")}, I.{Q("ID_INSCRIPTION")}, E.{Q("MATRICULE")}, E.{Q("NOM")}, E.{Q("PRENOM")}, " +
            $"N.{Q("VALEUR_NOTE")}, N.{Q("ABSENT")}, N.{Q("OBSERVATION")} " +
            $"FROM (({Q(Tables.Inscription)} AS I " +
            $"INNER JOIN {Q(Tables.Etudiant)} AS E ON I.{Q("ID_ETUDIANT")} = E.{Q("ID_ETUDIANT")}) " +
            $"LEFT JOIN {Q(Tables.Note)} AS N ON N.{Q("ID_INSCRIPTION")} = I.{Q("ID_INSCRIPTION")} AND N.{Q("ID_EVALUATION")} = ?) " +
            $"WHERE I.{Q("ID_CLASSE")} = ? ORDER BY E.{Q("NOM")}, E.{Q("PRENOM")}",
            MapNoteSaisieRow, P(idEvaluation), P(idClasse));

    public int CountNotesByEvaluation(int idEvaluation)
        => ScalarInt($"SELECT COUNT(*) FROM {Q(Tables.Note)} WHERE {Q("ID_EVALUATION")} = ?", P(idEvaluation));

    public int InsertNote(Note note)
        => InsertAndGetId(
            $"INSERT INTO {Q(Tables.Note)} ({Q("ID_EVALUATION")}, {Q("ID_INSCRIPTION")}, {Q("VALEUR_NOTE")}, " +
            $"{Q("ABSENT")}, {Q("OBSERVATION")}, {Q("DATE_SAISIE")}, {Q("CODE_UTR")}) VALUES (?, ?, ?, ?, ?, ?, ?)",
            P(note.IdEvaluation), P(note.IdInscription), P(note.ValeurNote), P(note.Absent ?? false),
            P(note.Observation), P(note.DateSaisie ?? DateTime.Now), P(note.CodeUtr));

    public int UpdateNote(Note note)
        => Execute(
            $"UPDATE {Q(Tables.Note)} SET {Q("VALEUR_NOTE")} = ?, {Q("ABSENT")} = ?, {Q("OBSERVATION")} = ?, " +
            $"{Q("DATE_SAISIE")} = ?, {Q("CODE_UTR")} = ? WHERE {Q("ID_NOTE")} = ?",
            P(note.ValeurNote), P(note.Absent ?? false), P(note.Observation),
            P(note.DateSaisie ?? DateTime.Now), P(note.CodeUtr), P(note.IdNote));

    public int DeleteNote(int idNote)
        => Execute($"DELETE FROM {Q(Tables.Note)} WHERE {Q("ID_NOTE")} = ?", P(idNote));

    /// <summary>Moyennes sur 20 par matière (requête R_MOYENNE_MATIERE, calcul vérifié).</summary>
    public List<MoyenneMatiereRow> ListMoyennesMatiere(int? idClasse = null, int? idPeriode = null, int? idInscription = null)
    {
        var conditions = new List<string>();
        var parameters = new List<System.Data.OleDb.OleDbParameter>();
        if (idClasse.HasValue) { conditions.Add(Q("ID_CLASSE") + " = ?"); parameters.Add(P(idClasse.Value)); }
        if (idPeriode.HasValue) { conditions.Add(Q("ID_PERIODE") + " = ?"); parameters.Add(P(idPeriode.Value)); }
        if (idInscription.HasValue) { conditions.Add(Q("ID_INSCRIPTION") + " = ?"); parameters.Add(P(idInscription.Value)); }
        var sql = $"SELECT * FROM {Q(SavedQueries.MoyenneMatiere)}";
        if (conditions.Count > 0) sql += " WHERE " + string.Join(" AND ", conditions);
        return QueryList(sql, MapMoyenneMatiere, parameters.ToArray());
    }

    /// <summary>Moyennes générales pondérées par période (requête R_MOYENNE_PERIODE).</summary>
    public List<MoyennePeriodeRow> ListMoyennesPeriode(int? idClasse = null, int? idPeriode = null)
    {
        var conditions = new List<string>();
        var parameters = new List<System.Data.OleDb.OleDbParameter>();
        if (idClasse.HasValue) { conditions.Add(Q("ID_CLASSE") + " = ?"); parameters.Add(P(idClasse.Value)); }
        if (idPeriode.HasValue) { conditions.Add(Q("ID_PERIODE") + " = ?"); parameters.Add(P(idPeriode.Value)); }
        var sql = $"SELECT * FROM {Q(SavedQueries.MoyennePeriode)}";
        if (conditions.Count > 0) sql += " WHERE " + string.Join(" AND ", conditions);
        return QueryList(sql, MapMoyennePeriode, parameters.ToArray());
    }

    // ----- Mappers -----

    public static PeriodeEval MapPeriode(DataRow row) => new()
    {
        IdPeriode = DataRowMapper.GetInt32(row, "ID_PERIODE"),
        IdAnnee = DataRowMapper.GetInt32(row, "ID_ANNEE"),
        CodePeriode = DataRowMapper.GetString(row, "CODE_PERIODE"),
        Libelle = DataRowMapper.GetString(row, "LIBELLE"),
        OrdrePer = DataRowMapper.GetInt32(row, "ORDRE_PER"),
        Ponderation = DataRowMapper.GetDouble(row, "PONDERATION"),
        DateDebut = DataRowMapper.GetDateTime(row, "DATE_DEBUT"),
        DateFin = DataRowMapper.GetDateTime(row, "DATE_FIN"),
        Cloturee = DataRowMapper.GetBoolean(row, "CLOTUREE")
    };

    public static Evaluation MapEvaluation(DataRow row) => new()
    {
        IdEvaluation = DataRowMapper.GetInt32(row, "ID_EVALUATION"),
        IdPeriode = DataRowMapper.GetInt32(row, "ID_PERIODE"),
        IdProg = DataRowMapper.GetInt32(row, "ID_PROG"),
        Intitule = DataRowMapper.GetString(row, "INTITULE"),
        Nature = DataRowMapper.GetString(row, "NATURE"),
        DateEval = DataRowMapper.GetDateTime(row, "DATE_EVAL"),
        Bareme = DataRowMapper.GetDouble(row, "BAREME"),
        Poids = DataRowMapper.GetDouble(row, "POIDS"),
        Publiee = DataRowMapper.GetBoolean(row, "PUBLIEE")
    };

    public static EvaluationDetail MapEvaluationDetail(DataRow row) => new()
    {
        IdEvaluation = DataRowMapper.GetInt32(row, "ID_EVALUATION"),
        IdPeriode = DataRowMapper.GetInt32(row, "ID_PERIODE"),
        Periode = DataRowMapper.GetString(row, "PERIODE_LIB"),
        IdProg = DataRowMapper.GetInt32(row, "ID_PROG"),
        IdClasse = DataRowMapper.GetInt32(row, "ID_CLASSE"),
        Classe = DataRowMapper.GetString(row, "CLASSE_LIB"),
        CodeMatiere = DataRowMapper.GetString(row, "CODE_MATIERE"),
        Matiere = DataRowMapper.GetString(row, "MATIERE_LIB"),
        Intitule = DataRowMapper.GetString(row, "INTITULE"),
        Nature = DataRowMapper.GetString(row, "NATURE"),
        DateEval = DataRowMapper.GetDateTime(row, "DATE_EVAL"),
        Bareme = DataRowMapper.GetDouble(row, "BAREME"),
        Poids = DataRowMapper.GetDouble(row, "POIDS"),
        Publiee = DataRowMapper.GetBoolean(row, "PUBLIEE"),
        PeriodeCloturee = DataRowMapper.GetBoolean(row, "PERIODE_CLOTUREE")
    };

    public static Note MapNote(DataRow row) => new()
    {
        IdNote = DataRowMapper.GetInt32(row, "ID_NOTE"),
        IdEvaluation = DataRowMapper.GetInt32(row, "ID_EVALUATION"),
        IdInscription = DataRowMapper.GetInt32(row, "ID_INSCRIPTION"),
        ValeurNote = DataRowMapper.GetDouble(row, "VALEUR_NOTE"),
        Absent = DataRowMapper.GetBoolean(row, "ABSENT"),
        Observation = DataRowMapper.GetString(row, "OBSERVATION"),
        DateSaisie = DataRowMapper.GetDateTime(row, "DATE_SAISIE"),
        CodeUtr = DataRowMapper.GetString(row, "CODE_UTR")
    };

    public static NoteSaisieRow MapNoteSaisieRow(DataRow row) => new()
    {
        IdNote = DataRowMapper.GetInt32(row, "ID_NOTE"),
        IdInscription = DataRowMapper.GetInt32(row, "ID_INSCRIPTION"),
        Matricule = DataRowMapper.GetString(row, "MATRICULE"),
        Nom = DataRowMapper.GetString(row, "NOM"),
        Prenom = DataRowMapper.GetString(row, "PRENOM"),
        ValeurNote = DataRowMapper.GetDouble(row, "VALEUR_NOTE"),
        Absent = DataRowMapper.GetBoolean(row, "ABSENT", false),
        Observation = DataRowMapper.GetString(row, "OBSERVATION")
    };

    public static MoyenneMatiereRow MapMoyenneMatiere(DataRow row) => new()
    {
        IdInscription = DataRowMapper.GetInt32(row, "ID_INSCRIPTION"),
        IdPeriode = DataRowMapper.GetInt32(row, "ID_PERIODE"),
        IdClasse = DataRowMapper.GetInt32(row, "ID_CLASSE"),
        CodeMatiere = DataRowMapper.GetString(row, "CODE_MATIERE"),
        Coefficient = DataRowMapper.GetDouble(row, "COEFFICIENT"),
        IdFormateur = DataRowMapper.GetInt32(row, "ID_FORMATEUR"),
        MoyenneMat = DataRowMapper.GetDouble(row, "MOYENNE_MAT")
    };

    public static MoyennePeriodeRow MapMoyennePeriode(DataRow row) => new()
    {
        IdInscription = DataRowMapper.GetInt32(row, "ID_INSCRIPTION"),
        IdPeriode = DataRowMapper.GetInt32(row, "ID_PERIODE"),
        IdClasse = DataRowMapper.GetInt32(row, "ID_CLASSE"),
        TotalPoints = DataRowMapper.GetDouble(row, "TOTAL_POINTS"),
        TotalCoef = DataRowMapper.GetDouble(row, "TOTAL_COEF"),
        Moyenne = DataRowMapper.GetDouble(row, "MOYENNE")
    };
}

File: ltadmin/repositories/evaluation_repository.py
"""Dépôt des évaluations et des notes de contrôle continu."""

from __future__ import annotations

from typing import List, Optional

from ltadmin.data import row_mapper as rm
from ltadmin.data.schema import SavedQueries, Tables
from ltadmin.models.dto import (
    EvaluationDetail,
    MoyenneMatiereRow,
    MoyennePeriodeRow,
    NoteSaisieRow,
)
from ltadmin.models.entities import Evaluation, Note, PeriodeEval
from ltadmin.repositories.base import RepositoryBase, p, q


class EvaluationRepository(RepositoryBase):

    # ----- PERIODE_EVAL -----

    def get_periode(self, id_periode: int) -> Optional[PeriodeEval]:
        return self.query_single(
            f"SELECT * FROM {q(Tables.PERIODE_EVAL)} WHERE {q('ID_PERIODE')} = ?",
            EvaluationRepository.map_periode, [p(id_periode)])

    def list_periodes(self, id_annee: Optional[int] = None) -> List[PeriodeEval]:
        if id_annee is not None:
            return self.query_list(
                f"SELECT * FROM {q(Tables.PERIODE_EVAL)} WHERE {q('ID_ANNEE')} = ? "
                f"ORDER BY {q('ORDRE_PER')}",
                EvaluationRepository.map_periode, [p(id_annee)])
        return self.query_list(
            f"SELECT * FROM {q(Tables.PERIODE_EVAL)} ORDER BY {q('ID_ANNEE')}, {q('ORDRE_PER')}",
            EvaluationRepository.map_periode)

    def insert_periode(self, periode: PeriodeEval) -> int:
        return self.insert_and_get_id(
            f"INSERT INTO {q(Tables.PERIODE_EVAL)} ({q('ID_ANNEE')}, {q('CODE_PERIODE')}, "
            f"{q('LIBELLE')}, {q('ORDRE_PER')}, {q('PONDERATION')}, {q('DATE_DEBUT')}, "
            f"{q('DATE_FIN')}, {q('CLOTUREE')}) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
            [p(periode.id_annee), p(periode.code_periode), p(periode.libelle),
             p(periode.ordre_per), p(periode.ponderation), p(periode.date_debut),
             p(periode.date_fin), p(False if periode.cloturee is None else periode.cloturee)])

    def update_periode(self, periode: PeriodeEval) -> int:
        return self.execute(
            f"UPDATE {q(Tables.PERIODE_EVAL)} SET {q('ID_ANNEE')} = ?, {q('CODE_PERIODE')} = ?, "
            f"{q('LIBELLE')} = ?, {q('ORDRE_PER')} = ?, {q('PONDERATION')} = ?, "
            f"{q('DATE_DEBUT')} = ?, {q('DATE_FIN')} = ?, {q('CLOTUREE')} = ? "
            f"WHERE {q('ID_PERIODE')} = ?",
            [p(periode.id_annee), p(periode.code_periode), p(periode.libelle),
             p(periode.ordre_per), p(periode.ponderation), p(periode.date_debut),
             p(periode.date_fin), p(False if periode.cloturee is None else periode.cloturee),
             p(periode.id_periode)])

    def set_periode_cloturee(self, id_periode: int, cloturee: bool) -> int:
        return self.execute(
            f"UPDATE {q(Tables.PERIODE_EVAL)} SET {q('CLOTUREE')} = ? WHERE {q('ID_PERIODE')} = ?",
            [p(cloturee), p(id_periode)])

    def delete_periode(self, id_periode: int) -> int:
        return self.execute(
            f"DELETE FROM {q(Tables.PERIODE_EVAL)} WHERE {q('ID_PERIODE')} = ?", [p(id_periode)])

    # ----- EVALUATION -----

    def get_evaluation(self, id_evaluation: int) -> Optional[Evaluation]:
        return self.query_single(
            f"SELECT * FROM {q(Tables.EVALUATION)} WHERE {q('ID_EVALUATION')} = ?",
            EvaluationRepository.map_evaluation, [p(id_evaluation)])

    def get_evaluation_detail(self, id_evaluation: int) -> Optional[EvaluationDetail]:
        return self.query_single(
            EvaluationRepository.evaluation_detail_sql() + f" WHERE EV.{q('ID_EVALUATION')} = ?",
            EvaluationRepository.map_evaluation_detail, [p(id_evaluation)])

    def list_evaluations(self, id_periode: Optional[int] = None,
                         id_classe: Optional[int] = None,
                         id_prog: Optional[int] = None) -> List[EvaluationDetail]:
        conditions = []
        parameters = []
        if id_periode is not None:
            conditions.append(f"EV.{q('ID_PERIODE')} = ?")
            parameters.append(p(id_periode))
        if id_classe is not None:
            conditions.append(f"P.{q('ID_CLASSE')} = ?")
            parameters.append(p(id_classe))
        if id_prog is not None:
            conditions.append(f"EV.{q('ID_PROG')} = ?")
            parameters.append(p(id_prog))
        sql = EvaluationRepository.evaluation_detail_sql()
        if conditions:
            sql += " WHERE " + " AND ".join(conditions)
        sql += f" ORDER BY EV.{q('DATE_EVAL')}, EV.{q('ID_EVALUATION')}"
        return self.query_list(sql, EvaluationRepository.map_evaluation_detail, parameters)

    @staticmethod
    def evaluation_detail_sql() -> str:
        return (
            f"SELECT EV.*, PE.{q('LIBELLE')} AS PERIODE_LIB, PE.{q('CLOTUREE')} AS PERIODE_CLOTUREE, "
            f"P.{q('ID_CLASSE')} AS ID_CLASSE, C.{q('LIBELLE')} AS CLASSE_LIB, "
            f"P.{q('CODE_MATIERE')} AS CODE_MATIERE, M.{q('MATIERE')} AS MATIERE_LIB "
            f"FROM (((({q(Tables.EVALUATION)} AS EV "
            f"INNER JOIN {q(Tables.PERIODE_EVAL)} AS PE ON EV.{q('ID_PERIODE')} = PE.{q('ID_PERIODE')}) "
            f"INNER JOIN {q(Tables.PROGRAMME)} AS P ON EV.{q('ID_PROG')} = P.{q('ID_PROG')}) "
            f"INNER JOIN {q(Tables.CLASSE)} AS C ON P.{q('ID_CLASSE')} = C.{q('ID_CLASSE')}) "
            f"INNER JOIN {q(Tables.MATIERE)} AS M ON P.{q('CODE_MATIERE')} = M.{q('CODE_MATIERE')})"
        )

    def insert_evaluation(self, evaluation: Evaluation) -> int:
        return self.insert_and_get_id(
            f"INSERT INTO {q(Tables.EVALUATION)} ({q('ID_PERIODE')}, {q('ID_PROG')}, "
            f"{q('INTITULE')}, {q('NATURE')}, {q('DATE_EVAL')}, {q('BAREME')}, {q('POIDS')}, "
            f"{q('PUBLIEE')}) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
            [p(evaluation.id_periode), p(evaluation.id_prog), p(evaluation.intitule),
             p(evaluation.nature), p(evaluation.date_eval), p(evaluation.bareme),
             p(evaluation.poids), p(False if evaluation.publiee is None else evaluation.publiee)])

    def update_evaluation(self, evaluation: Evaluation) -> int:
        return self.execute(
            f"UPDATE {q(Tables.EVALUATION)} SET {q('ID_PERIODE')} = ?, {q('ID_PROG')} = ?, "
            f"{q('INTITULE')} = ?, {q('NATURE')} = ?, {q('DATE_EVAL')} = ?, {q('BAREME')} = ?, "
            f"{q('POIDS')} = ?, {q('PUBLIEE')} = ? WHERE {q('ID_EVALUATION')} = ?",
            [p(evaluation.id_periode), p(evaluation.id_prog), p(evaluation.intitule),
             p(evaluation.nature), p(evaluation.date_eval), p(evaluation.bareme),
             p(evaluation.poids), p(False if evaluation.publiee is None else evaluation.publiee),
             p(evaluation.id_evaluation)])

    def set_evaluation_publiee(self, id_evaluation: int, publiee: bool) -> int:
        return self.execute(
            f"UPDATE {q(Tables.EVALUATION)} SET {q('PUBLIEE')} = ? WHERE {q('ID_EVALUATION')} = ?",
            [p(publiee), p(id_evaluation)])

    def delete_evaluation(self, id_evaluation: int) -> int:
        return self.execute(
            f"DELETE FROM {q(Tables.EVALUATION)} WHERE {q('ID_EVALUATION')} = ?",
            [p(id_evaluation)])

    # ----- NOTE -----

    def get_note(self, id_evaluation: int, id_inscription: int) -> Optional[Note]:
        return self.query_single(
            f"SELECT * FROM {q(Tables.NOTE)} WHERE {q('ID_EVALUATION')} = ? "
            f"AND {q('ID_INSCRIPTION')} = ?",
            EvaluationRepository.map_note, [p(id_evaluation), p(id_inscription)])

    def list_grille_saisie(self, id_evaluation: int, id_classe: int) -> List[NoteSaisieRow]:
        """Grille de saisie : tous les inscrits de la classe + leur note éventuelle."""
        return self.query_list(
            f"SELECT N.{q('ID_NOTE')}, I.{q('ID_INSCRIPTION')}, E.{q('MATRICULE')}, "
            f"E.{q('NOM')}, E.{q('PRENOM')}, N.{q('VALEUR_NOTE')}, N.{q('ABSENT')}, "
            f"N.{q('OBSERVATION')} "
            f"FROM (({q(Tables.INSCRIPTION)} AS I "
            f"INNER JOIN {q(Tables.ETUDIANT)} AS E ON I.{q('ID_ETUDIANT')} = E.{q('ID_ETUDIANT')}) "
            f"LEFT JOIN {q(Tables.NOTE)} AS N ON N.{q('ID_INSCRIPTION')} = I.{q('ID_INSCRIPTION')} "
            f"AND N.{q('ID_EVALUATION')} = ?) "
            f"WHERE I.{q('ID_CLASSE')} = ? ORDER BY E.{q('NOM')}, E.{q('PRENOM')}",
            EvaluationRepository.map_note_saisie_row, [p(id_evaluation), p(id_classe)])

    def count_notes_by_evaluation(self, id_evaluation: int) -> int:
        return self.scalar_int(
            f"SELECT COUNT(*) FROM {q(Tables.NOTE)} WHERE {q('ID_EVALUATION')} = ?",
            [p(id_evaluation)])

    def insert_note(self, note: Note) -> int:
        return self.insert_and_get_id(
            f"INSERT INTO {q(Tables.NOTE)} ({q('ID_EVALUATION')}, {q('ID_INSCRIPTION')}, "
            f"{q('VALEUR_NOTE')}, {q('ABSENT')}, {q('OBSERVATION')}, {q('DATE_SAISIE')}, "
            f"{q('CODE_UTR')}) VALUES (?, ?, ?, ?, ?, ?, ?)",
            [p(note.id_evaluation), p(note.id_inscription), p(note.valeur_note),
             p(False if note.absent is None else note.absent), p(note.observation),
             p(note.date_saisie or __import__("datetime").datetime.now()), p(note.code_utr)])

    def update_note(self, note: Note) -> int:
        return self.execute(
            f"UPDATE {q(Tables.NOTE)} SET {q('VALEUR_NOTE')} = ?, {q('ABSENT')} = ?, "
            f"{q('OBSERVATION')} = ?, {q('DATE_SAISIE')} = ?, {q('CODE_UTR')} = ? "
            f"WHERE {q('ID_NOTE')} = ?",
            [p(note.valeur_note), p(False if note.absent is None else note.absent),
             p(note.observation), p(note.date_saisie or __import__("datetime").datetime.now()),
             p(note.code_utr), p(note.id_note)])

    def delete_note(self, id_note: int) -> int:
        return self.execute(
            f"DELETE FROM {q(Tables.NOTE)} WHERE {q('ID_NOTE')} = ?", [p(id_note)])

    # ----- Moyennes (requêtes enregistrées) -----

    def list_moyennes_matiere(self, id_classe: Optional[int] = None,
                              id_periode: Optional[int] = None,
                              id_inscription: Optional[int] = None) -> List[MoyenneMatiereRow]:
        """Moyennes sur 20 par matière (requête R_MOYENNE_MATIERE)."""
        conditions = []
        parameters = []
        if id_classe is not None:
            conditions.append(f"{q('ID_CLASSE')} = ?")
            parameters.append(p(id_classe))
        if id_periode is not None:
            conditions.append(f"{q('ID_PERIODE')} = ?")
            parameters.append(p(id_periode))
        if id_inscription is not None:
            conditions.append(f"{q('ID_INSCRIPTION')} = ?")
            parameters.append(p(id_inscription))
        sql = f"SELECT * FROM {q(SavedQueries.MOYENNE_MATIERE)}"
        if conditions:
            sql += " WHERE " + " AND ".join(conditions)
        return self.query_list(sql, EvaluationRepository.map_moyenne_matiere, parameters)

    def list_moyennes_periode(self, id_classe: Optional[int] = None,
                              id_periode: Optional[int] = None) -> List[MoyennePeriodeRow]:
        """Moyennes générales pondérées par période (requête R_MOYENNE_PERIODE)."""
        conditions = []
        parameters = []
        if id_classe is not None:
            conditions.append(f"{q('ID_CLASSE')} = ?")
            parameters.append(p(id_classe))
        if id_periode is not None:
            conditions.append(f"{q('ID_PERIODE')} = ?")
            parameters.append(p(id_periode))
        sql = f"SELECT * FROM {q(SavedQueries.MOYENNE_PERIODE)}"
        if conditions:
            sql += " WHERE " + " AND ".join(conditions)
        return self.query_list(sql, EvaluationRepository.map_moyenne_periode, parameters)

    # ----- Mappers -----

    @staticmethod
    def map_periode(row: dict) -> PeriodeEval:
        return PeriodeEval(
            id_periode=rm.get_int(row, "ID_PERIODE"),
            id_annee=rm.get_int(row, "ID_ANNEE"),
            code_periode=rm.get_string(row, "CODE_PERIODE"),
            libelle=rm.get_string(row, "LIBELLE"),
            ordre_per=rm.get_int(row, "ORDRE_PER"),
            ponderation=rm.get_double(row, "PONDERATION"),
            date_debut=rm.get_datetime(row, "DATE_DEBUT"),
            date_fin=rm.get_datetime(row, "DATE_FIN"),
            cloturee=rm.get_bool(row, "CLOTUREE"),
        )

    @staticmethod
    def map_evaluation(row: dict) -> Evaluation:
        return Evaluation(
            id_evaluation=rm.get_int(row, "ID_EVALUATION"),
            id_periode=rm.get_int(row, "ID_PERIODE"),
            id_prog=rm.get_int(row, "ID_PROG"),
            intitule=rm.get_string(row, "INTITULE"),
            nature=rm.get_string(row, "NATURE"),
            date_eval=rm.get_datetime(row, "DATE_EVAL"),
            bareme=rm.get_double(row, "BAREME"),
            poids=rm.get_double(row, "POIDS"),
            publiee=rm.get_bool(row, "PUBLIEE"),
        )

    @staticmethod
    def map_evaluation_detail(row: dict) -> EvaluationDetail:
        return EvaluationDetail(
            id_evaluation=rm.get_int(row, "ID_EVALUATION"),
            id_periode=rm.get_int(row, "ID_PERIODE"),
            periode=rm.get_string(row, "PERIODE_LIB"),
            id_prog=rm.get_int(row, "ID_PROG"),
            id_classe=rm.get_int(row, "ID_CLASSE"),
            classe=rm.get_string(row, "CLASSE_LIB"),
            code_matiere=rm.get_string(row, "CODE_MATIERE"),
            matiere=rm.get_string(row, "MATIERE_LIB"),
            intitule=rm.get_string(row, "INTITULE"),
            nature=rm.get_string(row, "NATURE"),
            date_eval=rm.get_datetime(row, "DATE_EVAL"),
            bareme=rm.get_double(row, "BAREME"),
            poids=rm.get_double(row, "POIDS"),
            publiee=rm.get_bool(row, "PUBLIEE"),
            periode_cloturee=rm.get_bool(row, "PERIODE_CLOTUREE"),
        )

    @staticmethod
    def map_note(row: dict) -> Note:
        return Note(
            id_note=rm.get_int(row, "ID_NOTE"),
            id_evaluation=rm.get_int(row, "ID_EVALUATION"),
            id_inscription=rm.get_int(row, "ID_INSCRIPTION"),
            valeur_note=rm.get_double(row, "VALEUR_NOTE"),
            absent=rm.get_bool(row, "ABSENT"),
            observation=rm.get_string(row, "OBSERVATION"),
            date_saisie=rm.get_datetime(row, "DATE_SAISIE"),
            code_utr=rm.get_string(row, "CODE_UTR"),
        )

    @staticmethod
    def map_note_saisie_row(row: dict) -> NoteSaisieRow:
        return NoteSaisieRow(
            id_note=rm.get_int(row, "ID_NOTE"),
            id_inscription=rm.get_int(row, "ID_INSCRIPTION"),
            matricule=rm.get_string(row, "MATRICULE"),
            nom=rm.get_string(row, "NOM"),
            prenom=rm.get_string(row, "PRENOM"),
            valeur_note=rm.get_double(row, "VALEUR_NOTE"),
            absent=rm.get_bool_or(row, "ABSENT", False),
            observation=rm.get_string(row, "OBSERVATION"),
        )

    @staticmethod
    def map_moyenne_matiere(row: dict) -> MoyenneMatiereRow:
        return MoyenneMatiereRow(
            id_inscription=rm.get_int(row, "ID_INSCRIPTION"),
            id_periode=rm.get_int(row, "ID_PERIODE"),
            id_classe=rm.get_int(row, "ID_CLASSE"),
            code_matiere=rm.get_string(row, "CODE_MATIERE"),
            coefficient=rm.get_double(row, "COEFFICIENT"),
            id_formateur=rm.get_int(row, "ID_FORMATEUR"),
            moyenne_mat=rm.get_double(row, "MOYENNE_MAT"),
        )

    @staticmethod
    def map_moyenne_periode(row: dict) -> MoyennePeriodeRow:
        return MoyennePeriodeRow(
            id_inscription=rm.get_int(row, "ID_INSCRIPTION"),
            id_periode=rm.get_int(row, "ID_PERIODE"),
            id_classe=rm.get_int(row, "ID_CLASSE"),
            total_points=rm.get_double(row, "TOTAL_POINTS"),
            total_coef=rm.get_double(row, "TOTAL_COEF"),
            moyenne=rm.get_double(row, "MOYENNE"),
        )
